### ISSUE TITLE
extra-* settings are order depedendent: extra-foo option is a noop if used before foo in the configuration file

### ISSUE THREAD

**Issue by symphorien** [NixOS/nix#9487]
**Describe the bug**

`extra-substituters = foo` is ignored in /etc/nix/nix.conf if used before `substituters = ` but not after.

**Steps To Reproduce**

1. replace /etc/nix/nix.conf by a file if necessary
2. Add `extra-substituters = https://test` before `substituters = https://cache.nixos.org/`
3. the settings is ignored:
```
$  nix show-config | grep substituters
substituters = https://cache.nixos.org/
trusted-substituters = 
```

Compare to what happens when `extra-substituters = https://test` is after `substituters = https://cache.nixos.org/`:
```
$  nix show-config | grep substituters
substituters = https://cache.nixos.org/ https://test
trusted-substituters = 
```

**Expected behavior**

The effect of `extra-*` options is not order dependent: it should append to values affected to the main option even if the main option comes after.

**`nix-env --version` output**

reproduced with 2.13.6 and 2.18.1

**Additional context**

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by thufschmitt** [comment]
This is coherent with the overall semantics of the config files which are read imperatively from top to bottom (meaning that if you put twice the same setting item, the second one will shadow the first one).

This is arguably not optimal, but I don't think we'll want to touch it as it would be a very easy backwards-compatibility footgun

**Comment by Ericson2314** [comment]
Yeah the best time to revisit this is if we switch to JSON for a new config format, #8373.

**Comment by symphorien** [comment]
Can we get a warning then ? Just like passing --option substituters gives a warning when the user is not trusted instead of silently ignoring the setting.

**Comment by symphorien** [comment]
Workaround proposed downstream as https://github.com/NixOS/nixpkgs/pull/278064

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.

**Comment by Malix-Labs** [comment]
Is that not still relevant?

**Comment by symphorien** [comment]
I can still reproduce with nix 2.34.8 and there is still no warning.
Reading the list of linked PRs/issues above shows both NixOS and nix-darwin have code to work around this issue.
Even if the current semantics must be enshrined, I think https://github.com/NixOS/nix/issues/9487#issuecomment-1836704973 is still actionable.
For these reasons I think this issue deserves to be reopened.
